docs/madr/decisions/081-datasource-api.md

```diff
@@ -0,0 +1,256 @@
+
+* Resolved by the control plane (e.g., for signing certificates in `MeshIdentity`)
+* Accessed at runtime by the data plane (e.g., for mutual TLS in `MeshExternalService` on `ZoneEgress`)
+
+#### Pros
+* One unified structure
+* Can be reused consistently across different resources
+* Inline secrets are not synced to the global anymore
+
+#### Cons
+* Unclear who is responsible for delivering the configuration (control plane or ZoneEgress)
+* Not obvious where values like file paths or environment variables need to be set (control plane or data plane context)
+* Potential security risk: private keys might be delivered by the control plane
+
+
+### Create different DataSources based on where should be accessed
+
+```golang
+// +kubebuilder:validation:Enum=File;Secret;EnvVar
+type CPType string
+
+const (
+    FileType   CPType = "File"
+    SecretType CPType = "Secret"
+    EnvVarType CPType = "EnvVar"
+)
+
+type ControlPlaneDataSource struct {
+    Type      CPType       `json:"type,omitempty"`
+    File      *File      `json:"file,omitempty"`
+    SecretRef *SecretRef `json:"secretRef,omitempty"`
+    EnvVar    *EnvVar    `json:"envVar,omitempty"`
+}
+
+// +kubebuilder:validation:Enum=File;EnvVar
+type DpType string
+
+const (
+    FileType   DpType = "File"
+    EnvVarType DpType = "EnvVar"
+)
+
+type DataplaneDataSource struct {
+    Type      DpType       `json:"type,omitempty"`
+    File      *File      `json:"file,omitempty"`
+    EnvVar    *EnvVar    `json:"envVar,omitempty"`
+}
+
+type File struct {
+    Path string `json:"path,omitempty"`
+}
+
+// +kubebuilder:validation:Enum=Secret
+type RefType string
+
+const (
+    SecretRefType RefType = "Secret"
+)
+
+type SecretRef struct {
+    Kind RefType `json:"kind,omitempty"`
+    Name string  `json:"name,omitempty"`
+}
+
+type EnvVar struct {
+    Name string `json:"name,omitempty"`
+}
+```
+
+The proposed model addresses the issues found in the previous approach, as it clearly defines where the DataSource should be resolved (control plane or data plane). It also avoids exposing irrelevant options (e.g., using a Secret where it can’t be accessed by the data plane).
+
+#### Pros
+* Clear API boundaries — it’s explicitly defined what can be configured for each resource type.
+* Improved security — avoids delivering sensitive material like private keys to the data plane via SDS.
+* Inline secrets are not synced to the global anymore
+
+#### Cons
+* Two similar but separate structures — which adds some duplication and potential maintenance overhead.
+* No explicit support for Secret in data plane context — users would need to manually mount secrets as files or environment variables.
+* Lack of inline
+* Might be problem with rotation - Envoy might not read if it's rotated - might be support by: https://www.envoyproxy.io/docs/envoy/latest/configuration/security/secret#key-rotation
+
```

Comment:
Any alternative options?

Thinking about it:
Secret makes little sense on the DP (except if we use the serviceaccount token to retrieve the secret on the dp but that looks weird).
So we could for example have a resolveMode: dp|cp on the file and env var datasource?
We could then depending on the use allow only one mode?

I'd love us to come up with some alternatives and probably also clarify what is it we're trying to achieve here.

Reply:
Is this the same thing I'm suggesting here https://github.com/kumahq/kuma/pull/13910#discussion_r2196931897 ? 🤔